### ISSUE TITLE
DMD not required, should the README change?

### ISSUE THREAD

**Issue by russel** [jacob-carlborg/dstep#121]
I tried building DStep from Git master using Dub and it correctly finds ldc2 as the compiler, as I do not have DMD installed. The DStep README indicates you have to have DMD, but this is seemingly not the case. Should teh README be updated to indicate that Dub will find a D compiler if one is available?

**Comment by jacob-carlborg** [comment]
Not until ldc is officially supported. That is, it's regularly tested by adding it to .travis.yml.

**Comment by russel** [comment]
That's fair. I'll see if I can extend the Travis-CI matrix. I guess the only way of testing is to make a pull request?

**Comment by jacob-carlborg** [comment]
Yeah, to test on Travis-CI.

**Comment by jacob-carlborg** [comment]
There's another problem. I have  a branch, that should be merged eventually, that needs to run a configuration script written in D. That needs to be compiled, currently it just invokes DMD, but that won't work if LDC should be supported as well [1].

[1] https://github.com/jacob-carlborg/dstep/blob/static-linking-clang/configure#L1

**Comment by Dicebot** [comment]
> That needs to be compiled, currently it just invokes DMD, but that won't work if LDC should be supported as well [1]

Why wouldn't it? You should be able to both run it via `ldmd` and use inline dub projects to pick compiler automatically.

**Comment by jacob-carlborg** [comment]
> Why wouldn't it? You should be able to both run it via `ldmd`

What do you mean "both"?

> use inline dub projects to pick compiler automatically.

Ah, I didn't think of that. But what if I manually override the compiler with `dub build --compiler=`?

**Comment by Dicebot** [comment]
> What do you mean "both"?

That two options are present :)

> But what if I manually override the compiler with dub build --compiler=?

It will use the one explicitly specified, unless I misunderstand your question.

**Comment by jacob-carlborg** [comment]
> That two options are present :)

Ah, I didn't read that correct the first time. I still don't understand how `ldmd` will help.

> It will use the one explicitly specified, unless I misunderstand your question.

If I run `dub build` it will also compile and run the configuration script, which (currently) compiles using DMD. If I would to change the configuration script to use an inline Dub project I assume with would pick up the compiler correctly. But if my default compiler is DMD and I run `dub build --compiler=ldc`. Then when the configuration is built, how would it know to use `ldc`?

**Comment by jacob-carlborg** [comment]
@Dicebot This is not an issue, Dub already sets an environment variable `$DC` with the current compiler. It's just not documented [1].

[1] https://github.com/dlang/dub/issues/1073#issuecomment-279562840

**Comment by jacob-carlborg** [comment]
@russel When I added support for building with LDC, I encountered a segmentation fault that occurred on macOS with LDC but not on Linux in Travis CI using LDC. This is now fixed in master, but could you please checkout the following commit cc83cbc4523878ed6cec500e581d8e46534223db and try running DStep with the following command on Linux:

```
./bin/dstep test_files/clang-c/BuildSystem.h test_files/clang-c/CXCompilationDatabase.h test_files/clang-c/CXErrorCode.h test_files/clang-c/CXString.h test_files/clang-c/Documentation.h test_files/clang-c/Index.h test_files/clang-c/Platform.h -Itest_files --public-submodules --package clang.c -Iresources -o asd
```

I would like to know if it's failing on Linux as well, outside a Docker environment, when reporting the bug to LDC.

**Comment by russel** [comment]
I pulled DStep and checkout out cc83cbc4523878ed6cec500e581d8e46534223db . Using Debian Sid, I built with:
```
dub build --compiler=ldc2 --build=release
```
and then executed the line as requested:
```
|> ./bin/dstep test_files/clang-c/BuildSystem.h test_files/clang-c/CXCompilationDatabase.h test_files/clang-c/CXErrorCode.h test_files/clang-c/CXString.h test_files/clang-c/Documentation.h test_files/clang-c/Index.h test_files/clang-c/Platform.h -Itest_files --public-submodules --package clang.c -Iresources -o asd

|> 
```
so no error.

```
|> ldc2 --version
LDC - the LLVM D compiler (1.1.1):
  based on DMD v2.071.2 and LLVM 4.0.1
  built with LDC - the LLVM D compiler (0.17.4)
  Default target: x86_64-pc-linux-gnu
  Host CPU: penryn
  http://dlang.org - http://wiki.dlang.org/LDC
```

**Comment by jacob-carlborg** [comment]
Ok, thanks. Then it seems to be a macOS specific error.

**Comment by russel** [comment]
I can try things out on OS X El Capitan if that helps.

**Comment by jacob-carlborg** [comment]
No, I don't think that's necessary. I reproduced it on macOS both locally and on Travis CI. I've already added a workaround in DStep for the bug, by making the TLS variable a non-TLS immutable variable, which is actually more correct since it's read only.

I was just asking about Linux because it could help when I report the bug to LDC, which I've already done https://github.com/ldc-developers/ldc/issues/2187.
